loadbalancer/algorithm_test.go

```diff
@@ -411,8 +412,54 @@ func TestConsistentHashBoundedLoadDistribution(t *testing.T) {
 	}
 }
 
+func TestConsistentHashKeyDistribution(t *testing.T) {
+	endpoints := []string{"http://10.2.0.1:8080", "http://10.2.0.2:8080", "http://10.2.0.3:8080", "http://10.2.0.4:8080", "http://10.2.0.5:8080", "http://10.2.0.6:8080", "http://10.2.0.7:8080", "http://10.2.0.8:8080", "http://10.2.0.9:8080", "http://10.2.0.10:8080"}
+
+	stdDev1hashPerEndpoint := measureStdDev(t, endpoints, 1)
+	stdDev100HashesPerEndpoint := measureStdDev(t, endpoints, 100)
+
+	if stdDev100HashesPerEndpoint >= stdDev1hashPerEndpoint {
+		t.Errorf("Standard deviation with 100 hashes per endpoint should be lower than with 1 hash per endpoint. 100 hashes: %f, 1 hash: %f", stdDev100HashesPerEndpoint, stdDev1hashPerEndpoint)
+	}
+
+	if stdDev100HashesPerEndpoint >= 10 { // arbitrary target to flag accidental breaking changes. Currently is 5.93
+		t.Errorf("Standard deviation was too high for 100 vnodes, got %f", stdDev100HashesPerEndpoint)
+	}
+}
+
 func addInflightRequests(endpoint routing.LBEndpoint, count int) {
 	for i := 0; i < count; i++ {
 		endpoint.Metrics.IncInflightRequest()
 	}
 }
+
+// Measures how fair the hash ring is to each endpoint.
+// i.e. Of the possible hashes, how many will go to each endpoint. Lower the standard deviation the better.
+func measureStdDev(t *testing.T, endpoints []string, hashesPerEndpoint int) float64 {
+	ch := newConsistentHashInternal(endpoints, hashesPerEndpoint).(consistentHash)
+	ringOwnership := map[int]uint64{}
+	prevPartitionEndHash := uint64(0)
+	for i := 0; i < len(ch); i++ {
+		endpointIndex := ch[i].index
+		partitionEndHash := uint64(ch[i].hash)
+		ringOwnership[endpointIndex] += partitionEndHash - prevPartitionEndHash
+		prevPartitionEndHash = partitionEndHash
+	}
+	ringOwnership[0] += math.MaxUint64 - prevPartitionEndHash
```

Comment:
`ringOwnership[ch[0].index]` shouldn't it be?

Reply:
Yeah it should, good catch